lib/events/session_writer.go

```diff
@@ -281,17 +259,21 @@ func (a *AuditWriter) maybeSetBackoff(backoffUntil time.Time) bool {
 	}
 }
 
-// EmitAuditEvent emits audit event
-func (a *AuditWriter) EmitAuditEvent(ctx context.Context, event apievents.AuditEvent) error {
-	// Event modification is done under lock and in the same goroutine
-	// as the caller to avoid data races and event copying
-	if err := a.setupEvent(event); err != nil {
+// RecordEvent emits audit event
+func (a *SessionWriter) RecordEvent(ctx context.Context, event apievents.AuditEvent) error {
+	if err := checkBasicEventFields(event); err != nil {
 		return trace.Wrap(err)
 	}
 
+	// the index starts at 0, so we'll only know if the index is invalid
+	// if we've seen at least one event already
+	if event.GetIndex() == 0 && a.acceptedEvents.Load() > 0 {
+		return trace.BadParameter("missing mandatory event index field")
+	}
```

Comment:
I assume no op setter will never be used with SessionWriter?

Reply:
Yep a no-op setter should only be used for tests or when events will be discarded